File: services/social_ts/src/session.ts
import type { FastifyRequest } from "fastify";
import type { Redis } from "ioredis";

export type SessionUser = {
  version: number;
  user_id: string;
  username: string;
  role: string;
  is_active: boolean;
  created_at?: string;
};

const SESSION_UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
const RFC3339 = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})$/;

function validSessionShape(value: SessionUser): boolean {
  const version = value.version;
  if (version !== 1) return false;
  if (!value.user_id || !value.username || value.username.length > 100) return false;
  if (value.role !== "user" && value.role !== "admin") return false;
  if (value.is_active !== true) return false;
  if (!SESSION_UUID.test(value.user_id)) return false;
  if (!value.created_at || !RFC3339.test(value.created_at) || Number.isNaN(Date.parse(value.created_at))) return false;
  return true;
}

export async function sessionFromRequest(
  request: FastifyRequest,
  redis: Redis,
  cookieName: string,
): Promise<SessionUser | null> {
  const sessionId = request.cookies[cookieName];
  if (!sessionId) return null;

  const raw = await redis.get(`session:${sessionId}`);
  if (!raw) return null;

  try {
    const value = JSON.parse(raw) as SessionUser;
    return validSessionShape(value) ? value : null;
  } catch {
    return null;
  }
}

export async function requireUser(
  request: FastifyRequest,
  redis: Redis,
  cookieName: string,
): Promise<SessionUser> {
  const user = await sessionFromRequest(request, redis, cookieName);
  if (!user) {
    const err = new Error("Not authenticated.") as Error & { statusCode?: number };
    err.statusCode = 401;
    throw err;
  }
  return user;
}

export async function requireAdmin(
  request: FastifyRequest,
  redis: Redis,
  cookieName: string,
): Promise<SessionUser> {
  const user = await requireUser(request, redis, cookieName);
  if (user.role !== "admin") {
    const err = new Error("Admin access required.") as Error & { statusCode?: number };
    err.statusCode = 403;
    throw err;
  }
  return user;
}

File: services/catalog_go/internal/session/session.go
package session

import (
	"context"
	"encoding/json"
	"net/http"
	"regexp"
	"time"

	"github.com/redis/go-redis/v9"
)

type Service struct {
	redis      *redis.Client
	cookieName string
}

type Session struct {
	Version   int    `json:"version"`
	UserID    string `json:"user_id"`
	Username  string `json:"username"`
	Role      string `json:"role"`
	IsActive  bool   `json:"is_active"`
	CreatedAt string `json:"created_at"`
}

var sessionUUID = regexp.MustCompile(`^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$`)

func validV1SessionFields(s Session) bool {
	if !sessionUUID.MatchString(s.UserID) {
		return false
	}
	if s.CreatedAt == "" {
		return false
	}
	_, err := time.Parse(time.RFC3339Nano, s.CreatedAt)
	return err == nil
}

func (s Session) Valid() bool {
	// Session contract v1 is mandatory in the consolidated baseline.
	if s.Version != 1 {
		return false
	}
	if s.UserID == "" || s.Username == "" || len(s.Username) > 100 || !s.IsActive {
		return false
	}
	if s.Role != "user" && s.Role != "admin" {
		return false
	}
	if !validV1SessionFields(s) {
		return false
	}
	return true
}

func New(client *redis.Client, cookieName string) *Service {
	if cookieName == "" {
		cookieName = "session_id"
	}
	return &Service{redis: client, cookieName: cookieName}
}

func (s *Service) OptionalFromRequest(ctx context.Context, r *http.Request) *Session {
	cookie, err := r.Cookie(s.cookieName)
	if err != nil || cookie.Value == "" {
		return nil
	}
	raw, err := s.redis.Get(ctx, "session:"+cookie.Value).Result()
	if err != nil {
		return nil
	}
	var sess Session
	if json.Unmarshal([]byte(raw), &sess) != nil || !sess.Valid() {
		return nil
	}
	return &sess
}

func (s *Service) RequireAdmin(ctx context.Context, r *http.Request) (*Session, int, string) {
	sess := s.OptionalFromRequest(ctx, r)
	if sess == nil {
		return nil, http.StatusUnauthorized, "Not authenticated"
	}
	if sess.Role != "admin" {
		return nil, http.StatusForbidden, "Admin access required"
	}
	return sess, 0, ""
}
